src/nvim/fileio.c

```diff
@@ -2698,21 +2697,57 @@ int vim_rename(const char *from, const char *to)
   }
 
   // Rename() failed, try copying the file.
+  int ret = vim_copyfile(from, to);
+  if (ret != OK) {
+    return -1;
+  }
+
+  if (os_fileinfo(from, &from_info)) {
+    os_remove(from);
+  }
+
+  return 0;
+}
+
+/// Create the new file with same permissions as the original.
+/// Return -1 for failure, 0 for success.
+int vim_copyfile(const char *from, const char *to)
+{
+  char *errmsg = NULL;
+
+#ifdef HAVE_READLINK
+  FileInfo from_info;
+  if (os_fileinfo_link(from, &from_info) && S_ISLNK(from_info.stat.st_mode)) {
+    int ret = FAIL;
+
+    char linkbuf[MAXPATHL + 1];
+    ssize_t len = readlink(from, linkbuf, MAXPATHL);
+    if (len > 0) {
+      linkbuf[len] = NUL;
+
+      // Create link
+      ret = symlink(linkbuf, to);
+    }
+
+    return ret == 0 ? OK : FAIL;
+  }
+#endif
+
   int perm = os_getperm(from);
   // For systems that support ACL: get the ACL from the original file.
   vim_acl_T acl = os_get_acl(from);
   int fd_in = os_open(from, O_RDONLY, 0);
   if (fd_in < 0) {
     os_free_acl(acl);
-    return -1;
+    return FAIL;
   }
 
   // Create the new file with same permissions as the original.
   int fd_out = os_open(to, O_CREAT|O_EXCL|O_WRONLY|O_NOFOLLOW, perm);
```

Comment:
Do we gain anything if we use `os_copy(..., UV_FS_COPYFILE_FICLONE)` ?

https://github.com/neovim/neovim/blob/8df6736ca14d09f87cf0a8486758ac5708819434/src/nvim/bufwrite.c#L1715

Reply:
I'm not sure if `UV_FS_COPYFILE_FICLONE` should be used here, but `UV_FS_COPYFILE_EXCL` is needed. With `UV_FS_COPYFILE_EXCL`, `os_copy()` does seems to be a drop-in replacement here, and is faster and handles permissions better.